src/QuorumTools/Control.hs

```diff
@@ -52,35 +47,33 @@ find' predicate = Fold step Nothing id where
 event :: forall m e. MonadManaged m => e -> m (Async e, IO ())
 event e = do
   mvar <- liftIO newEmptyMVar
-  occurred <- awaitMVar mvar
+  occurred <- fork $ takeMVar mvar
   let fire = void $ tryPutMVar mvar e
   pure (occurred, fire)
 
--- | Creates a stream of values for a single publisher to update, and many
+-- | Creates a stream of values for a many publishers to update, and many
 -- consumers to subscribe (to value changes) or observe the current value.
---
--- NOTE: if we used STM+TChan+TMVar, we could remove the single-producer
--- restriction by modifying the chan and mvar concurrently.
---
 behavior :: (MonadIO m, Monoid a) => m (Behavior a)
-behavior = Behavior <$> liftIO newChan <*> liftIO (newMVar mempty)
+behavior = liftIO $ atomically $ Behavior <$> newTChan <*> newTMVar mempty
 
 transition :: MonadIO m => Behavior a -> (a -> a) -> m ()
-transition (Behavior chan mvar) update = liftIO $ do
-  nextValue <- pureModifyMVar mvar update
-  writeChan chan nextValue
+transition (Behavior tc tm) f = liftIO $ atomically $ do
+  prev <- takeTMVar tm
+  let next = f prev
+  putTMVar tm next
+  writeTChan tc next
 
-subscribe :: MonadIO m => Behavior a -> m (Chan a)
-subscribe (Behavior chan _) = liftIO $ dupChan chan
+subscribe :: MonadIO m => Behavior a -> m (TChan a)
+subscribe (Behavior chan _) = liftIO $ atomically $ dupTChan chan
 
 observe :: MonadIO m => Behavior a -> m a
-observe (Behavior _ mvar) = liftIO $ readMVar mvar
+observe (Behavior _ mvar) = liftIO $ atomically $ readTMVar mvar
 
 --
 -- TODO: this won't leak memory via the duplicated channel after our event
 -- occurs, right?
 --
 watch :: MonadManaged m => Behavior a -> (a -> Maybe b) -> m (Async b)
-watch b check = do
+watch b decide = do
   chan <- subscribe b
-  fork $ liftIO $ untilJust $ check <$> readChan chan
+  fork $ liftIO $ untilJust $ decide <$> atomically (readTChan chan)
```

Comment:
Yeah, it's necessary -- without `atomically`, you only have `STM a`. `atomically` runs the `STM a` action to produce an `IO a`.